src/network_asio.cpp

```diff
@@ -109,28 +114,91 @@ void connection::handle_connect(const boost::system::error_code& ec, endpoint en
 #else
 		LOG_NW << "Connected to " << endpoint->endpoint().address() << '\n';
 #endif
+		if(endpoint.address().is_loopback()) {
+			use_tls_ = false;
+		}
 		handshake();
 	}
 }
 
 void connection::handshake()
 {
 	static const uint32_t handshake = 0;
+	static const uint32_t tls_handshake = htonl(uint32_t(1));
 
-	boost::asio::async_write(socket_, boost::asio::buffer(reinterpret_cast<const char*>(&handshake), 4),
-		std::bind(&connection::handle_write, this, std::placeholders::_1, std::placeholders::_2));
+	boost::asio::async_write(
+		utils::get<raw_socket>(socket_),
+		boost::asio::buffer(use_tls_ ? reinterpret_cast<const char*>(&tls_handshake) : reinterpret_cast<const char*>(&handshake), 4),
+		std::bind(&connection::handle_write, this, std::placeholders::_1, std::placeholders::_2)
+	);
 
-	boost::asio::async_read(socket_, boost::asio::buffer(&handshake_response_.binary, 4),
+	boost::asio::async_read(utils::get<raw_socket>(socket_), boost::asio::buffer(&handshake_response_.binary, 4),
 		std::bind(&connection::handle_handshake, this, std::placeholders::_1));
 }
 
 void connection::handle_handshake(const boost::system::error_code& ec)
 {
 	if(ec) {
+		if(ec == boost::asio::error::eof && use_tls_) {
+			// immediate disconnect likely means old server not supporting TLS handshake code
+			fallback_to_unencrypted();
+			return;
+		}
+
 		throw system_error(ec);
 	}
+	
+	if(use_tls_) {
+		if(handshake_response_.num == 0xFFFFFFFFU) {
+			use_tls_ = false;
+			handle_handshake(ec);
+			return;
+		}
 
-	done_ = true;
+		if(handshake_response_.num == 0x00000000) {
+			tls_context_.set_default_verify_paths();
+			raw_socket s { std::move(utils::get<raw_socket>(socket_)) };
+			socket_.emplace<1>(std::move(s), tls_context_);
+			
+			auto& socket { utils::get<tls_socket>(socket_) };
+
+			socket.set_verify_mode(
+				boost::asio::ssl::verify_peer |
+				boost::asio::ssl::verify_fail_if_no_peer_cert
+			);
+
+#if BOOST_VERSION >= 107300
+			socket.set_verify_callback(boost::asio::ssl::host_name_verification(host_));
+#else
+			socket.set_verify_callback(boost::asio::ssl::rfc2818_verification(host_));
+#endif
+
+			socket.async_handshake(boost::asio::ssl::stream_base::client, [this](const boost::system::error_code& ec) {
+				if(ec) {
+					throw system_error(ec);
+				}
+
+				done_ = true;
+			});
+			return;
+		}
+
+		fallback_to_unencrypted();
+	} else {
+		done_ = true;
+	}
+}
+
+void connection::fallback_to_unencrypted()
+{
+	assert(use_tls_ == true);
+	use_tls_ = false;
+
+	boost::asio::ip::tcp::endpoint endpoint { utils::get<raw_socket>(socket_).remote_endpoint() };
```

Comment:
How come you don't use `visit` here like you do in other places (and on the next few lines)?

Reply:
socket_ is guaranteed to always have raw_socket at this point.

Reply:
Right, but you can still use a visitor. You just happen to know what the type passed in will be.

Reply:
But get is simpler. And visitor just handle both cases generating unneeded extra code for tls branch.